README.md
node-openzwave-shared
=====================
[Linux![Linux Build Status](https://travis-ci.org/OpenZWave/node-openzwave-shared.svg?branch=master)](https://travis-ci.org/OpenZWave/node-openzwave-shared)
[Windows![Windows Build status](https://ci.appveyor.com/api/projects/status/txg360huomtpgc8o?svg=true)](https://ci.appveyor.com/project/ekarak/node-openzwave-shared)

![Linux Build status](https://github.com/OpenZWave/node-openzwave-shared/workflows/Linux%20Builds/badge.svg)

![Windows Build status](https://github.com/OpenZWave/node-openzwave-shared/workflows/Windows%20Builds/badge.svg)

[![Join the chat at https://gitter.im/OpenZWave/node-openzwave-shared](https://badges.gitter.im/OpenZWave/node-openzwave-shared.svg)](https://gitter.im/OpenZWave/node-openzwave-shared?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Google group](http://www.google.com/images/icons/product/groups-32.png)](https://groups.google.com/d/forum/node-openzwave-shared)

This is the homepage for *node-openzwave-shared*, the official binary add-on for
Node.js, which wraps [Open ZWave](http://openzwave.com/), a high quality C++
library for controlling Z-Wave networks.

<img src="https://nodejs.org/static/images/logos/nodejs.png" style="width: 200px"/>
<img src="http://www.openzwave.com/css/ozwlogo.png"/>

You can now easily control *and manage* your [ZWave](http://www.z-wave.com/)
devices (lights, dimmers, blinds, you name it) from within [Node.js](https://nodejs.org/)
applications. [This library also supports *secure* devices](../master/README-security.md)
(eg door locks) that require encryption.

All widely used Node.js versions are supported with the help of [NaN](https://github.com/nodejs/nan).
This add-on is currently tested against 0.12.x on ARM and 4.2.x on x86_64, but
it should also work on all supported Node.js versions (as early as 0.11.12) and
architectures.

Check out the [Node-Red integration project](https://github.com/OpenZWave/node-red-contrib-openzwave)
for an interesting use case, in which I've wired up ZWave and KNX devices
working together as one big happy automated home.

This addon is currently able to:
- *scan* a Z-Wave network and report on connected devices,
- *write* values to zwave nodes
- *monitor* the network for changes,
- *heal* nodes and/or the network
- *perform* management tasks (add/remove nodes, replace failed nodes, manage
	their group associations etc)

**API change notice (v1.4.0)**
`enablePoll()` and `disablePoll()` used to affect *only the first ValueID of any given command class*.
This is *wrong*, as for multi-instance devices you probably need to poll multiple ValueID's.
These calls now accept a valueId, in very much the same way as `setValue()` does. This means that
you now have to pass a valueID object (or its 4 constituents) for each value you want to enable/disable
polling for.


**Important notice**

This library differs from its [ancestor library](https://github.com/jperkin/node-openzwave)
in that it links *dynamically* to an OpenZWave **shared library**
by means of your system dynamic linker. This is  in contrast to
*statically* linking OpenZWave as part of the node.js addon.

Thus you need to have OpenZWave fully installed on your system (both the
compiled library AND the development headers) before trying to install this
little baby.
I know this diverges from the dominant npm paradigm, but with the shared lib approach:
  - compilation / installation is a lot faster and
  - OZW *minor* upgrades / bugfixes are way lot easier.

This also means that you need to be careful if you upgrade your
OZW library: you might need to rebuild this addon, otherwise you'd might
get api mismatch exceptions.

## Prerequisites

### Linux

You will need to ensure the OpenZWave library and headers are
installed first. You can do this one of two ways.

- Manually compiling OpenZwave involves the usual dance of 1) downloading its source tarball
([latest code from GitHub](https://github.com/OpenZWave/open-zwave/archive/master.zip) or
[a stable release](https://github.com/OpenZWave/open-zwave/releases)) or
from [the OpenZWave snapshots repository](http://old.openzwave.com/snapshots/)
and then 2) compiling it and installing on your system (`make && sudo make install && sudo ldconfig`)

- You could also install OpenZWave via a [precompiled package that's suitable for
your Linux distribution and architecture](http://old.openzwave.com/downloads/).
**Notice:** Be sure to install *BOTH the binary (libopenzwave-x.y) AND the
development package (libopenzwave-dev).*

### MacOS/X

The only dependency that you need before compiling is `pkg-config`. Then, you
could either pull down the OZW repo from Github and do a `make && sudo make install` ,
or you could try installing OpenZWave using `brew install open-zwave`.

### Windows

Since there is no standard installation location for Open Z-Wave on Windows, it
will be automatically downloaded, compiled, and installed when you install this module.

## Installation and test script

Whenever you have OpenZWave installed in your machine, then all you need to do is:

```
$ npm install openzwave-shared
```

To try it out, boot up NodeJS, and use the `.load` shell helper function to boot up a basic OpenZWave CLI. This will initialise and expose 1) a `zwave` object that you can use to send commands and 2) the `nodes` object to get a list of all nodes:

```js
$ node
> .load test2.js
...
...

// the 1st node is the USB controller stick
> console.log(nodes[1])
{ manufacturer: 'Aeotec',
  manufacturerid: '0x0086',
  product: 'Z-Stick S2',
  producttype: '0x0002',
  productid: '0x0001',
  type: 'Static PC Controller',
  name: '',
  loc: '',
  classes: { '32': { '0': [Object] } },
  ready: true }

// set dimmer (node 5) to 50%
> zwave.setValue(5,38,1,0,50);
undefined
> node5: changed: 38:Level:54->54
```

**Notice 1:** If you receive the error `cannot find -lopenzwave` on a 64-bit Linux
system, `libopenzwave.so` was likely compiled into `/usr/local/lib64`.
Run the terminal command `ld -lopenzwave --verbose` for a list of search
locations used. You can workaround this by providing a symlink to one of the
listed locations such as `/usr/local/lib`.
Run `sudo ln -s /usr/local/lib64/libopenzwave.so /usr/local/lib/libopenzwave.so`
creates symlink so that the file appears to be in the location that `ld` looks
in. Now `npm install` should work.

**Notice 2:** nodejs Debian/Raspbian package (v0.10.29) causes an issue REPLACE_INVALID_UTF8 (cf. https://github.com/nodejs/nan/issues/414). You can use node from [herokuapp](http://node-arm.herokuapp.com/) or upgrade some node packages: `sudo npm -g install npm node-gyp`.

In case you want to develop your application in TypeScript, there is also a TypeScript
declaration file available. Besides adding typechecking to your codebase this will also
help you to navigate the API and see the type of parameters available for each function. You can
download the
[latest version of the declaration file from GitHub](https://github.com/jbaron/openzwave-shared-tsd).

## Development documentation

- [Basic API usage](../master/README-api.md)
- [List of supported events](../master/README-events.md)
- [Management of nodes](../master/README-mgmt.md)
- [Security API](../master/README-security.md)
- [Usage example](../master/README-example.md)
- [Deprecated API calls in OpenZWave 1.6](../master/README-deprecated-16.md)
  
## Environment-specific documentation
- [Installing on Rapsbian](../master/README-raspbian.md)
- [Installing on Ubuntu 15.04](../master/README-ubuntu.md)

## License

The [Open Z-Wave](http://www.openzwave.com/) library that this
module heavily relies upon is licensed under the Lesser GPLv3.

Everything else (all the bits that I and Jonathan have written)
is under the vastly more sensible ISC license.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.7.2](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.7.1...v1.7.2)

> 3 November 2020

- fix: typo in utils.cc ManuacturerSpecificDBReady [`#391`](https://github.com/OpenZWave/node-openzwave-shared/pull/391)
- fix: typescript types [`#384`](https://github.com/OpenZWave/node-openzwave-shared/pull/384)
- docs: typo in url [`#382`](https://github.com/OpenZWave/node-openzwave-shared/pull/382)
- feat: github actions for ci [`#381`](https://github.com/OpenZWave/node-openzwave-shared/pull/381)

#### [v1.7.1](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.7.0...v1.7.1)

> 9 June 2020

- fix: compile errors [`b3c4ea9`](https://github.com/OpenZWave/node-openzwave-shared/commit/b3c4ea9ed88cbfc05c2b822bab75769dfb0e6a20)
- fix: use OZWManagerAssign to handle exceptions [`89b9872`](https://github.com/OpenZWave/node-openzwave-shared/commit/89b9872e014b71d0c83f1c61dd6d741a3f86a82f)
- Release 1.7.1 [`d87b204`](https://github.com/OpenZWave/node-openzwave-shared/commit/d87b20487b6ec774a02731aab5760f84d6a34891)

#### [v1.7.0](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.6.2...v1.7.0)

> 15 May 2020

- fix: update typescript annotations [`#376`](https://github.com/OpenZWave/node-openzwave-shared/pull/376)
- fix: linting of types declaration file [`#367`](https://github.com/OpenZWave/node-openzwave-shared/pull/367)
- fix: misspelling in OpenZWave options [`#369`](https://github.com/OpenZWave/node-openzwave-shared/pull/369)
- feat: expose GetNodeClassInformation [`#374`](https://github.com/OpenZWave/node-openzwave-shared/pull/374)
- Add support for building on Android [`#372`](https://github.com/OpenZWave/node-openzwave-shared/pull/372)
- pull from node-openzwave-shared master [`#1`](https://github.com/OpenZWave/node-openzwave-shared/pull/1)
- test appveyor CI with Node 10 [`#373`](https://github.com/OpenZWave/node-openzwave-shared/pull/373)
- Release 1.7.0 [`3e0ce8a`](https://github.com/OpenZWave/node-openzwave-shared/commit/3e0ce8aef645aa2fcc257380be6a1687b573d980)
- fix: test3.js [`66c461b`](https://github.com/OpenZWave/node-openzwave-shared/commit/66c461b0f6f29a584d741a5e58f80e03cd5fc8d1)

#### [v1.6.2](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.6.1...v1.6.2)

> 3 March 2020

- fix: Empty config and user path when calling updateOptions [`d867ec6`](https://github.com/OpenZWave/node-openzwave-shared/commit/d867ec608ad130b219f7006ba6b3a1838217211c)
- Revert "BREAKING CHANGE: List valueId values array becomes an Object value: valueString" [`e9c9ab4`](https://github.com/OpenZWave/node-openzwave-shared/commit/e9c9ab47a13423f59749840f477053a8fe1490d3)
- BREAKING CHANGE: List valueId values array becomes an Object value: valueString [`60272ab`](https://github.com/OpenZWave/node-openzwave-shared/commit/60272ab90bc829965160b5d0722ee8a126a5ce4f)
- Release 1.6.2 [`1161317`](https://github.com/OpenZWave/node-openzwave-shared/commit/1161317320b9534654b1115bf9a50f37de3d127f)
- Add ozwcache to ignored files [`f44e2e4`](https://github.com/OpenZWave/node-openzwave-shared/commit/f44e2e4a2ac5c5b343177e22638cd064b1473f43)

#### [v1.6.1](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.6.0...v1.6.1)

> 12 February 2020

- Add support for building on FreeBSD [`#366`](https://github.com/OpenZWave/node-openzwave-shared/pull/366)
- Release 1.6.1 [`db3e197`](https://github.com/OpenZWave/node-openzwave-shared/commit/db3e197e6b999d98b04e4e232f7ad5aa9075df64)

#### [v1.6.0](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.5.12...v1.6.0)

> 11 February 2020

- Changes to make the windows build compatible with Node 12. [`#364`](https://github.com/OpenZWave/node-openzwave-shared/pull/364)
- Release 1.6.0 [`460f0a8`](https://github.com/OpenZWave/node-openzwave-shared/commit/460f0a8e467278b38d4baed5c3b86c69611d5f1b)
- docs: Updated events [`58c4d9b`](https://github.com/OpenZWave/node-openzwave-shared/commit/58c4d9bc6bf17953da2635050887a742c3e70b8b)
- Change because node 0.12 doesn't support string interpolation. [`1df30cf`](https://github.com/OpenZWave/node-openzwave-shared/commit/1df30cf347c1f4e64d5542a6b9026cbf1a9d651f)
- Fix to be compatible with old old node. [`c4885f7`](https://github.com/OpenZWave/node-openzwave-shared/commit/c4885f73b69652bd0a0033f7038181e0ec780d98)

#### [v1.5.12](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.5.11...v1.5.12)

> 16 January 2020

- fix: Fixed 'connected' callback with ozw version [`fae401c`](https://github.com/OpenZWave/node-openzwave-shared/commit/fae401cea1e940c836ae971868849bca5476cd6d)
- feat: Added getOzwVersion method [`80f9dbf`](https://github.com/OpenZWave/node-openzwave-shared/commit/80f9dbfadb6d786b7549b08c120a038b3e53ad3d)
- Release 1.5.12 [`1e4a78a`](https://github.com/OpenZWave/node-openzwave-shared/commit/1e4a78a31ffcb0607ea757fc4576d97c8b7e5c48)

#### [v1.5.11](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.5.10...v1.5.11)

> 16 January 2020

- Fix build issue #356 [`#359`](https://github.com/OpenZWave/node-openzwave-shared/pull/359)
- Bugfix/issue#183 disconnect deadlock [`#329`](https://github.com/OpenZWave/node-openzwave-shared/pull/329)
- Adding missing typing for removeFailedNode [`#357`](https://github.com/OpenZWave/node-openzwave-shared/pull/357)
- Release 1.5.11 [`4a09346`](https://github.com/OpenZWave/node-openzwave-shared/commit/4a09346640bed0c450c38227733528241381144d)
- Adding missing typeing for removeFailedNode [`9635576`](https://github.com/OpenZWave/node-openzwave-shared/commit/9635576107e4ac2b19cf2f32c392c3aa87fa2d17)
- fix build [`52f7792`](https://github.com/OpenZWave/node-openzwave-shared/commit/52f7792c7de7d6fd74fc5d56aa97a1e1b6e50d29)

#### [v1.5.10](https://github.com/OpenZWave/node-openzwave-shared/compare/v1.5.9...v1.5.10)

> 5 November 2019


test2.js
var ZWave = require('./lib/openzwave-shared.js');
var os = require('os');

var zwave = new ZWave({
  ConsoleOutput: true
});

zwavedriverpaths = {
  "darwin": '/dev/cu.usbmodem1411',
  "linux": '/dev/ttyACM0',
  "windows": '\\\\.\\COM3'
}

var nodes = [];
var homeid = null;

zwave.on('connected', function (version) {
  console.log("**** CONNECTED ****")
  console.log("Openzwave version:", version)
});

zwave.on('driver ready', function (home_id) {
  homeid = home_id;
  console.log('scanning homeid=0x%s...', homeid.toString(16));
});

zwave.on('driver failed', function () {
  console.log('failed to start driver');
  process.exit();
});

zwave.on('node added', function (nodeid) {
  nodes[nodeid] = {
    manufacturer: '',
    manufacturerid: '',
    product: '',
    producttype: '',
    productid: '',
    type: '',
    name: '',
    loc: '',
    classes: {},
    ready: false,
  };
});

zwave.on('node event', function (nodeid, data) {
  console.log('node%d event: Basic set %d', nodeid, data);
});

zwave.on('value added', function (nodeid, comclass, value) {
  if (!nodes[nodeid]['classes'][comclass])
    nodes[nodeid]['classes'][comclass] = {};
  nodes[nodeid]['classes'][comclass][value.index] = value;
});

zwave.on('value changed', function (nodeid, comclass, value) {
  if (nodes[nodeid]['ready']) {
    console.log('node%d: changed: %d:%s:%s->%s', nodeid, comclass,
      value['label'],
      nodes[nodeid]['classes'][comclass][value.index]['value'],
      value['value']);
  }
  nodes[nodeid]['classes'][comclass][value.index] = value;
});

zwave.on('value removed', function (nodeid, comclass, index) {
  if (nodes[nodeid]['classes'][comclass] &&
    nodes[nodeid]['classes'][comclass][index])
    delete nodes[nodeid]['classes'][comclass][index];
});

zwave.on('node ready', function (nodeid, nodeinfo) {
  nodes[nodeid]['manufacturer'] = nodeinfo.manufacturer;
  nodes[nodeid]['manufacturerid'] = nodeinfo.manufacturerid;
  nodes[nodeid]['product'] = nodeinfo.product;
  nodes[nodeid]['producttype'] = nodeinfo.producttype;
  nodes[nodeid]['productid'] = nodeinfo.productid;
  nodes[nodeid]['type'] = nodeinfo.type;
  nodes[nodeid]['name'] = nodeinfo.name;
  nodes[nodeid]['loc'] = nodeinfo.loc;
  nodes[nodeid]['ready'] = true;
  console.log('node%d: %s, %s', nodeid,
    nodeinfo.manufacturer ? nodeinfo.manufacturer : 'id=' + nodeinfo.manufacturerid,
    nodeinfo.product ? nodeinfo.product : 'product=' + nodeinfo.productid +
      ', type=' + nodeinfo.producttype);
  console.log('node%d: name="%s", type="%s", location="%s"', nodeid,
    nodeinfo.name,
    nodeinfo.type,
    nodeinfo.loc);
  for (comclass in nodes[nodeid]['classes']) {
    switch (comclass) {
      case 0x25: // COMMAND_CLASS_SWITCH_BINARY
      case 0x26: // COMMAND_CLASS_SWITCH_MULTILEVEL
        zwave.enablePoll(nodeid, comclass);
        break;
    }
    var values = nodes[nodeid]['classes'][comclass];
    console.log('node%d: class %d', nodeid, comclass);
    for (idx in values)
      console.log('node%d:   %s=%s', nodeid, values[idx]['label'], values[
        idx]['value']);
  }
});

zwave.on('notification', function (nodeid, notif) {
  switch (notif) {
    case 0:
      console.log('node%d: message complete', nodeid);
      break;
    case 1:
      console.log('node%d: timeout', nodeid);
      break;
    case 2:
      console.log('node%d: nop', nodeid);
      break;
    case 3:
      console.log('node%d: node awake', nodeid);
      break;
    case 4:
      console.log('node%d: node sleep', nodeid);
      break;
    case 5:
      console.log('node%d: node dead', nodeid);
      break;
    case 6:
      console.log('node%d: node alive', nodeid);
      break;
  }
});

zwave.on('scan complete', function () {
  console.log('====> scan complete');
  // set dimmer node 5 to 50%
  //    zwave.setValue(5,38,1,0,50);
  //zwave.setValue({node_id:5,	class_id: 38,	instance:1,	index:0}, 50 );
  zwave.requestAllConfigParams(3);
});

zwave.on('controller command', function (n, rv, st, msg) {
  console.log(
    'controller commmand feedback: %s node==%d, retval=%d, state=%d', msg,
    n, rv, st);
});

console.log("connecting to " + zwavedriverpaths[os.platform()]);
zwave.connect(zwavedriverpaths[os.platform()]);

process.on('SIGINT', function () {
  console.log('disconnecting...');
  zwave.disconnect(zwavedriverpaths[os.platform()]);
  process.exit();
});
